### PR TITLE
[webview_flutter] Support for handling basic authentication requests

### PR DESCRIPTION
**Pull request by andreisas06**
### Description
This pull request introduces the handling of HTTP Basic Authentication in our WebView implementation. Specifically, it includes the following changes:

1. Added setHttpAuthCredentials() function: This function allows developers to programmatically set credentials (username and password) for a given host and realm.
2. Implemented onReceivedHttpAuthRequest() callback: This callback is invoked when the WebView encounters a website that requires HTTP Basic Authentication. The host and realm information from the authentication request is utilized to match and use the stored credentials.

### Issues fixed by PR:
Closes flutter/flutter#83556

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

**Comment by andreisas06** [comment]
@bparrishMines should I proceed with the PR containing only the platform interface package?

File: packages/webview_flutter/webview_flutter_platform_interface/lib/src/platform_navigation_delegate.dart
Comment (bparrishMines) [inline packages/webview_flutter/webview_flutter_platform_interface/lib/src/platform_navigation_delegate.dart:33]: nit: 
```suggestion
/// Signature for callbacks that report basic auth events by the native web view.
```

File: packages/webview_flutter/webview_flutter_platform_interface/lib/src/platform_webview_controller.dart
Comment (bparrishMines) [inline packages/webview_flutter/webview_flutter_platform_interface/lib/src/platform_webview_controller.dart:264]: nit:
```suggestion
  /// Sets the auth credentials for basic authentication and can be used with
  /// [PlatformNavigationDelegate.onReceiveHttpAuthRequest].
```

File: packages/webview_flutter/webview_flutter_android/android/src/main/java/io/flutter/plugins/webviewflutter/WebViewClientHostApiImpl.java
Comment (bparrishMines) [inline packages/webview_flutter/webview_flutter_android/android/src/main/java/io/flutter/plugins/webviewflutter/WebViewClientHostApiImpl.java:103]: The FlutterAPI should pass back the `HttpAuthHandler` instead of adding a `setAuthCredentials` method. Mainly because the `HttpAuthHandler` has multiple methods that could be used. e.g. `cancel` and [useHttpAuthUsernamePassword](https://developer.android.com/reference/android/webkit/HttpAuthHandler#useHttpAuthUsernamePassword()).

**Review by bparrishMines — COMMENTED** [review]
This also needs an implementation for `webview_flutter_wkwebview`. Using https://developer.apple.com/documentation/webkit/wknavigationdelegate/3601237-webview?language=objc I believe.

**Comment by andreisas06** [comment]
Hey @bparrishMines , @JeroenWeener will be taking over for the IOS implementation of this PR.

File: packages/webview_flutter/webview_flutter_wkwebview/lib/src/webkit_webview_controller.dart
Comment (bparrishMines) [inline packages/webview_flutter/webview_flutter_wkwebview/lib/src/webkit_webview_controller.dart:835]: Why does this only support `HTTPBasic`? I'm not that familiar with authentication, but would other methods not work with the response to the completionHandler?

File: packages/webview_flutter/webview_flutter_platform_interface/lib/src/types/http_basic_auth_request.dart
Comment (bparrishMines) [inline packages/webview_flutter/webview_flutter_platform_interface/lib/src/types/http_basic_auth_request.dart:16]: It looks like iOS can include more than just the `username` and `password` when proceeding (e.g. `NSURLCrendtial.persistence`). I think we can change the method to take a single parameter, something like `WebViewCredential`, and add the username and password to this class. Then we can add additional parameters with an iOS implementation in a future update.

**Review by bparrishMines — COMMENTED** [review]
Thanks for working on this. I added comments for the first review.

Also the Java API implementations still needs tests. See https://github.com/flutter/packages/tree/main/packages/webview_flutter/webview_flutter_android/android/src/test/java/io/flutter/plugins/webviewflutter

File: packages/webview_flutter/webview_flutter_platform_interface/lib/src/platform_navigation_delegate.dart
Comment (JeroenWeener) [reply packages/webview_flutter/webview_flutter_platform_interface/lib/src/platform_navigation_delegate.dart:141]: It is indeed because I only implemented basic HTTP support. I changed the name to `setOnHttpAuthRequest`, so we can use it for other auth types too.

File: packages/webview_flutter/webview_flutter_wkwebview/lib/src/webkit_webview_controller.dart
Comment (JeroenWeener) [reply packages/webview_flutter/webview_flutter_wkwebview/lib/src/webkit_webview_controller.dart:835]: I am also not too familiar with the other authentication methods. Instead of making a mistake here, my line of thinking was to take the conservative approach and limit functionality here. In the future, other authentication methods could be added.

File: packages/webview_flutter/webview_flutter_wkwebview/example/integration_test/webview_flutter_test.dart
Comment (JeroenWeener) [reply packages/webview_flutter/webview_flutter_wkwebview/example/integration_test/webview_flutter_test.dart:1193]: The test is indeed intended to only validate the happy path. The mock backend at line 37 does not check the credentials.

**Comment by JeroenWeener** [comment]
@bparrishMines I will add the required Java tests tomorrow

**Comment by JeroenWeener** [comment]
@bparrishMines Java tests are up!

File: packages/webview_flutter/webview_flutter_android/lib/src/android_webview.dart
Comment (bparrishMines) [inline packages/webview_flutter/webview_flutter_android/lib/src/android_webview.dart:1410]: The parameters for this method should stay `String username, String password`, since it is a mirror of the Android SDK. Unlike the platform interface, It doesn't need to be adaptable.

File: packages/webview_flutter/webview_flutter/lib/src/navigation_delegate.dart
Comment (bparrishMines) [inline packages/webview_flutter/webview_flutter/lib/src/navigation_delegate.dart:179]: This field can be removed due to https://github.com/flutter/packages/pull/4735#issuecomment-1688957524. But keep it as a constructor parameter though.

File: packages/webview_flutter/webview_flutter/example/lib/main.dart
Comment (bparrishMines) [inline packages/webview_flutter/webview_flutter/example/lib/main.dart:538]: This example looks great! The only problem is that is uses a third-party website. I think we could actually create our own HTML for this example.

**Review by bparrishMines — COMMENTED** [review]
This looks pretty much ready! I added a few more comments.

@stuartmorgan I think this is ready for a secondary review. 

Once the merge conflicts and nits are handled, I think we can go ahead and create the separate PR with just the platform interface changes.

Also, feel free to let me know if you would like me to take over the PR at any moment since I was delayed with the review.

File: packages/webview_flutter/webview_flutter/example/lib/main.dart
Comment (stuartmorgan) [reply packages/webview_flutter/webview_flutter/example/lib/main.dart:538]: HTTP authentication requires a server, not just HTML, which makes it problematic for an example app. For an integration test we can use a local test server, but I don't see a good solution here for controlling an endpoint.

Maybe we should allow pasting in a URL to load for this example, instead of hard-coding something, since I can't find an example page I'd want to put in our app.

File: packages/webview_flutter/webview_flutter/example/lib/main.dart
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter/example/lib/main.dart:18]: Why do we need this? It feels like we're not re-exporting something we should be if this is necessary for basic usage.

File: packages/webview_flutter/webview_flutter/example/pubspec.yaml
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter/example/pubspec.yaml:21]: Same here; I would not expect us to need this.

File: packages/webview_flutter/webview_flutter/example/integration_test/webview_flutter_test.dart
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter/example/integration_test/webview_flutter_test.dart:39]: We should probably test that we actually got the expected values. E.g., this test would pass even if an implementation accidentally swapped the username and password values.

File: packages/webview_flutter/webview_flutter_android/example/lib/main.dart
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter_android/example/lib/main.dart:132]: Does this add any value given that the dialog will include the same information?

File: packages/webview_flutter/webview_flutter_wkwebview/example/ios/Runner/Info.plist
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter_wkwebview/example/ios/Runner/Info.plist:36]: Why is this being added?

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

File: packages/webview_flutter/webview_flutter/example/lib/main.dart
Comment (JeroenWeener) [reply packages/webview_flutter/webview_flutter/example/lib/main.dart:538]: @stuartmorgan could you elaborate on your proposed solution? Are you suggesting to change the example app so it starts with a textfield element, allowing the user to navigate to any page they desire?

File: packages/webview_flutter/webview_flutter_wkwebview/example/ios/Runner/Info.plist
Comment (stuartmorgan) [reply packages/webview_flutter/webview_flutter_wkwebview/example/ios/Runner/Info.plist:36]: Where are you seeing `UIBackgroundModes` in the previous version of the file?

File: packages/webview_flutter/webview_flutter_platform_interface/lib/src/types/http_auth_request.dart
Comment (JeroenWeener) [reply packages/webview_flutter/webview_flutter_platform_interface/lib/src/types/http_auth_request.dart]: I went ahead and changed the code as suggested. Let me know if you want to see this changed.

Comment (bparrishMines) [reply packages/webview_flutter/webview_flutter_platform_interface/lib/src/types/http_auth_request.dart]: The iOS class has multiple responses to an authentication challenge: https://developer.apple.com/documentation/foundation/nsurlsessionauthchallengedisposition. I believe the rationality was that iOS could add more later if it is a method.

**Review by stuartmorgan — COMMENTED** [review]
Sorry, I accidentally had responses in draft form :(

**Comment by mvanbeusekom** [comment]
Hi @stuartmorgan, 

This PR is ready for review, we are keeping it up to date until then.

File: packages/webview_flutter/webview_flutter/example/lib/main.dart
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter/example/lib/main.dart:132]: Bad merge, or was this intentionally removed?

File: packages/webview_flutter/webview_flutter_platform_interface/lib/src/types/http_auth_request.dart
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter_platform_interface/lib/src/types/http_auth_request.dart:61]: I think the outcome of https://github.com/flutter/packages/pull/4140#discussion_r1320308186 was that we were going to go back to `onProceed` and `onCancel`? I'll defer to @bparrishMines on the exact callback structure though.

**Review by stuartmorgan — COMMENTED** [review]
I didn't review the implementations, but I'm happy with the overall structure. Once @bparrishMines has done another full review of the PR it should be ready to break up.

File: packages/webview_flutter/webview_flutter_platform_interface/lib/src/types/http_auth_request.dart
Comment (bparrishMines) [reply packages/webview_flutter/webview_flutter_platform_interface/lib/src/types/http_auth_request.dart:61]: Correct. I think we should separate this into  `onCancel` and `onProceed` again. And then iOS can add its additional method.

**Comment by bparrishMines** [comment]
I went ahead and added the changes mentioned in https://github.com/flutter/packages/pull/4140#discussion_r1380293879 and updated the PR with the pigeon upgrade from https://github.com/flutter/packages/pull/5271

**Review by bparrishMines — APPROVED** [review]
LGTM

I think this is ready to be split into separate PRs. @andreisas06 @mvanbeusekom let me know if you would like to do this or have me finish this part.

**Comment by JeroenWeener** [comment]
Hi @bparrishMines, thank you for the changes and the review.

Could you tell me how you usually deal with splitting PRs? I find myself a bit lost when trying to separate the platform interface changes from the other changes.

**Comment by stuartmorgan** [comment]
> Could you tell me how you usually deal with splitting PRs? I find myself a bit lost when trying to separate the platform interface changes from the other changes.

FWIW my workflow is:
- Make a new local branch from the combo PR branch (e.g., `git checkout -t webview-auth-request -b webview-auth-request-interface`)
- Merge in `main` so it's up to date
- Revert the other packages to the state on main (`git checkout main -- packages/webview_flutter packages/webview_flutter_android packages/webview_flutter_wkwebview; git commit`)

That should given you a branch that is exactly the platform interface subset of this PR. (I prefer this to trying to apply a patch file to a new clean branch because I find that more error-prone.)

**Comment by JeroenWeener** [comment]
Thanks @stuartmorgan, learned something new today 👍

**Comment by JeroenWeener** [comment]
@bparrishMines I created the platform interface PR. It can be found here: https://github.com/flutter/packages/pull/5362.

**Comment by JeroenWeener** [comment]
@stuartmorgan I celebrated too soon. Some changes that were made to the repo and were merged into this branch are also removed following the approach you suggested. For example, `webview_flutter/CHANGELOG.md` has its newest entries removed.

**Comment by JeroenWeener** [comment]
Thank @stuartmorgan and @bparrishMines for merging the related PRs. The final PR that exposes the iOS and Android implementations can be found here: https://github.com/flutter/packages/pull/5727.

@andreisas06 could you close this PR?

**Comment by jmagman** [comment]
@bparrishMines and @stuartmorgan can this be closed in favor of https://github.com/flutter/packages/pull/5727#pullrequestreview-1837479599 and friends?  Seems like this is zombied.

**Pull request closed without merging**
